datafusion/physical-expr/src/planner.rs

```diff
@@ -51,20 +51,23 @@ pub fn create_physical_expr(
     input_schema: &Schema,
     execution_props: &ExecutionProps,
 ) -> Result<Arc<dyn PhysicalExpr>> {
-    if input_dfschema.field_with_name(None, "_id").is_ok() {
-        if input_schema.fields.len() != input_dfschema.fields().len() {
+    if input_schema.fields.len() != input_dfschema.fields().len() {
+        // If `input_schema` and `input_dfschema` get different schemas, then
+        // there is one case where such a inconsistency is allowed, i.e., the
+        // `input_schema` has an extra `_id` field, which is added by us.
+
+        // Is this inconsistency made by us, if not, panic
+        let made_by_us = input_schema.field_with_name("_id").is_ok()
+            && input_dfschema.field_with_name(None, "_id").is_err();
```

Comment:
mostly LGTM, but I don't like we hard coded `_id` here, if we mount existing parquet files, there is a chance to have conflicts with existing schema.

Reply:
#2 has already enforced that the underlying Parquet files must have an `_id` field, should we revert that as well? i.e., when tweaking the generated physical plan, only add an `_id` field if it is actually present in the Parquet file, if they don't have any `_id` field, the documents are not treated as deleted

Something to note, given that this is a hack to DataFusion, it is not that easy to make it a general implementation (having an interface where one can specify the field used for deletion at runtime)

Reply:
Just make the `_id` field configurable, allow to specify which field will be treated as the key field when we mount a general Parquet datasource, in our's case, it is `_id` field, I don't think there is a issue for our hack, we did't change the data strucute under the hood.

Reply:
I am OK with merge this PR first, to keep it more flexiable can be another PR, it is low priority, before we support mount external datasource.

Reply:
> Just make the _id field configurable, allow to specify which field will be treated as the key field when we mount a general Parquet datasource

I agree this is better

>  I don't think there is a issue for our hack, we did't change the data strucute under the hood.

Maybe we can store it in the `SessionContext`, and make it configurable through a public interface

> I am OK with merge this PR first, to keep it more flexiable can be another PR, it is low priority, before we support mount external datasource.

Ok, let's merge this PR as-is then